Dummy PR to trigger review comments

ggaurav10 opened the pull request:
Dummy PR to trigger review comments

Commit:
Dummy PR to trigger review comments

amshuman-kr commented on kubernetes/autoscaling_v1alpha1_hvpa.yaml:
```diff
@@ -0,0 +1,52 @@
+apiVersion: autoscaling.k8s.io/v1alpha1
+kind: Hvpa
+metadata:
+  # Uncomment below annotation if you want kube-controller-manager to act on HPA's recommendation instead of hvpa-controller
```
Might be better to clean this up and introduce it only when we have enhanced kcm?

amshuman-kr commented on pkg/apis/autoscaling/v1alpha1/hvpa_types.go:
```diff
@@ -0,0 +1,204 @@
+
+	// ScaleUpDelay defines the minimum delay in minutes between 2 consecutive scale up operations
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h"
+	ScaleUpDelay string `json:"scaleUpDelay,omitempty"`
```
```suggestion
	ScaleUpStabilizationWindow *string `json:"scaleUpDelay,omitempty"`
```

amshuman-kr commented on pkg/apis/autoscaling/v1alpha1/hvpa_types.go:
```diff
@@ -0,0 +1,204 @@
+
+	// ScaleDownDelay defines the minimum delay in minutes between 2 consecutive scale down operations
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h"
+	ScaleDownDelay string `json:"scaleDownDelay,omitempty"`
```
```suggestion
	ScaleDownStabilizationWindow *string `json:"scaleDownDelay,omitempty"`
```

amshuman-kr commented on pkg/apis/autoscaling/v1alpha1/hvpa_types.go:
```diff
@@ -0,0 +1,204 @@
+type ChangeThreshold struct {
+	// Value is the absolute value of the threshold
+	// +optional
+	Value string `json:"value,omitempty"`
```
```suggestion
	Value *string `json:"value,omitempty"`
```

amshuman-kr commented on pkg/apis/autoscaling/v1alpha1/hvpa_types.go:
```diff
@@ -0,0 +1,204 @@
+	Value string `json:"value,omitempty"`
+	// Percentage is the percentage of currently allocated value to be used as threshold
+	// +optional
+	Percentage int32 `json:"percentage,omitempty"`
```
```suggestion
	Percentage *int32 `json:"percentage,omitempty"`
```

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+	lastFraction := autoscalingv1alpha1.VpaWeight(1)
+	lookupNextFraction := false
+
+	for _, interval := range hvpa.Spec.WeightBasedScalingIntervals {
```
If this for-loop is to pick the right interval/bucket it might be better to wrap this in a function.

ggaurav10 replied:
done

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+
+func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment, hpaWeight autoscalingv1alpha1.VpaWeight) (int32, error) {
+	anno := hvpa.GetAnnotations()
+	if val, ok := anno["hpa-controller"]; ok {
```
See the comments [here](https://github.com/ggaurav10/hvpa-controller/pull/5/files#diff-f181d9d7ccc73f657c9c60ab37a9bfd1R4).

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+		return weightedReplicas, err
+	}
+
+	log.Info("HPA", "no scaling done. Current replicas", currentReplicas)
```
Might be good to log (or even a condition in status) why scaling was not done. Because recommendations did not change or because the stabilisation window was effective?

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+	if thresholdVals == nil {
+		if resourceType == corev1.ResourceMemory {
+			// Default to 200M
+			quantity, _ = resource.ParseQuantity("200M")
```
Constants :-)

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+		}
+		if resourceType == corev1.ResourceCPU {
+			// Default to 200m
+			quantity, _ = resource.ParseQuantity("200m")
```
Constants :-)

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+	return status, nil
+}
+
+func (r *ReconcileHvpa) hpaScaleIfRequired(desiredReplicas, currentReplicas int32, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment, hpaWeight autoscalingv1alpha1.VpaWeight) (int32, error) {
```
Is this used?

ggaurav10 replied:
removed

amshuman-kr commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+	return 0, nil
+}
+
+func (r *ReconcileHvpa) vpaScaleIfRequired(recommendations *vpa_api.RecommendedPodResources, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment, vpaWeight autoscalingv1alpha1.VpaWeight) (bool, error) {
```
Is this used?

ggaurav10 replied:
removed

amshuman-kr reviewed (changes requested):
Some minor comments. Looks good otherwise.

ggaurav10 commented on pkg/controller/hvpa/hvpa_controller.go:
```diff
@@ -0,0 +1,651 @@
+	newDeploy := deployment.DeepCopy()
+	for _, rec := range recommendations.ContainerRecommendations {
+		for id, container := range newDeploy.Spec.Template.Spec.Containers {
+			// Currently only scale up is implemented, and vpaWeight is assumed to be 1
```
Need to remove this comment. It is not valid anymore

Commit:
Addressing review comments

Commit:
Instead of ceil when scaling down horizontally, consider floor value of the weighted numbber of replicas

Commit:
Requeue object when returning.
There can be a case, when hvpa applies partial recommendation because of fractional vpaWeight.
In this case, if there are no more events on watched events, the cluster will be stuck in a state when HPA and VPA still have recommendations, however,
 hvpa doesn't apply them.

Commit:
Adding files: CODEOWNERS, CONTRIBUTING.md, LICENSE.md, VERSION
